### ISSUE TITLE
Why do deduction guides for take_view and drop_view have different constraints? LWG 3447

### ISSUE THREAD

**Issue by cpplearner** [cplusplus/draft#3997]
In [\[range.take.view\]](https://eel.is/c++draft/range.take.view), the deduction guide for `take_view` is declared as:

```
  template<range R>
    take_view(R&&, range_difference_t<R>)
      -> take_view<views::all_t<R>>;
```

In [\[range.drop.view\]](https://eel.is/c++draft/range.drop.view), the deduction guide for `drop_view` is declared as:

```
  template<class R>
    drop_view(R&&, range_difference_t<R>) -> drop_view<views::all_t<R>>;
```

Note the difference between their template parameter lists.

AFAIK there's no difference in effect, because `views::all_t` only accepts a `viewable_range`.

Can they be declared in a more similar way?

**Comment by jensmaurer** [comment]
I believe the presence or absence (and the spelling) of associated constraints influences partial ordering, before the shape of the return value is even considered. So, this does not appear to be an editorial change.

@CaseyCarter , @jwakely ?

**Comment by jwakely** [comment]
`drop_view` was added later by [P1035R7](https://wg21.link/P1035R7), so the inconsistency could just be an oversight, but I agree it's not editorial.

**Comment by jwakely** [comment]
It looks like that `take_view` guide and this one are the only constrained guides for view adaptors:
```
template<input_range R>
    split_view(R&&, range_value_t<R>)
      -> split_view<views::all_t<R>, single_view<range_value_t<R>>>;
```

**Comment by CaseyCarter** [comment]
`take_view` significantly predates `drop_view`. When we added `take_view`, the `range` constraint was necessary to ensure that `iterator_t<R>` was well-formed in `iter_difference_t<iterator_t<R>>`.

P1035 constrains `iterator_t` itself with `range`, and added `range_difference_t<R>` as a `range`-constrained alias for `iter_difference_t<iterator_t<R>>` so neither `drop_view` nor `take_view` require this constraint. P1035 had blanket editorial instructions to replace `iter_difference_t<iterator_t<R>>` with `range_difference_t<R>`, but it didn't occur to the authors to also audit constraints elsewhere and remove any we'd made extraneous.

TLDR: Changing this would have no normative effect, but it falls into that grey area where we need experts to attest to that. I suggest filing an LWG issue.

**Comment by jensmaurer** [comment]
Handled by [LWG 3447](https://cplusplus.github.io/LWG/issue3447).

**Comment by johelegp** [comment]
> Status: WP

This can be closed.
